<a href="https://colab.research.google.com/github/helloohowareyu/Praktikum-Kecerdasan-Artifisial-Lanjut/blob/main/BAB12_ACHMAD%20YUSUF%20HAMDANI%20FIRMANSYAH_245150207111084.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# LAPORAN PRAKTIKUM KECERDASAN ARTIFISIAL LANJUT
**Nama: Achmad Yusuf Hamdani Firmansyah**

**NIM: 245150207111094**

**Kelas: KAL - E**

---
## Bab 12 : Association Rule


## Prosedur Percobaan

In [1]:
# Mengunduh library mlxtend

!pip install mlxtend

In [6]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [5]:
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [10]:
# Mensimulasikan dataset ritel
data = {
    'TransactionID': [1, 1, 2, 2, 3, 3, 3, 4, 4, 4],
    'Item': ['A', 'B', 'A', 'B', 'A', 'B', 'C', 'A', 'B', 'C']
}

df = pd.DataFrame(data)
df

,TransactionID,Item
0,1,A
1,1,B
2,2,A
3,2,B
4,3,A
5,3,B
6,3,C
7,4,A
8,4,B
9,4,C


In [12]:
# Melakukan one-hot encoding untuk mempersiapkan data untuk analisis market basket.

basket = pd.crosstab(index=df['TransactionID'], columns=df['Item']).map(lambda x: 1 if x > 0 else 0)
basket

Item,A,B,C
TransactionID,,,
1,1,1,0
2,1,1,0
3,1,1,1
4,1,1,1


In [13]:
# Menerapkan algoritma apriori untuk menemukan himpunan item yang sering muncul.
frequent_itemsets = apriori(basket, min_support=0.2, use_colnames=True)
frequent_itemsets

,support,itemsets
0,1.0,(A)
1,1.0,(B)
2,0.5,(C)
3,1.0,"(B, A)"
4,0.5,"(C, A)"
5,0.5,"(B, C)"
6,0.5,"(B, C, A)"


In [14]:
# Menghasilkan aturan asosiasi
rules = association_rules(frequent_itemsets, metric='lift', min_threshold=1)
rules

/usr/local/lib/python3.12/dist-packages/mlxtend/frequent_patterns/association_rules.py:186: RuntimeWarning: invalid value encountered in divide
  cert_metric = np.where(certainty_denom == 0, 0, certainty_num / certainty_denom)


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(B),(A),1.0,1.0,1.0,1.0,1.0,1.0,0.0,inf,0.0,1.0,0.0,1.00
1,(A),(B),1.0,1.0,1.0,1.0,1.0,1.0,0.0,inf,0.0,1.0,0.0,1.00
2,(C),(A),0.5,1.0,0.5,1.0,1.0,1.0,0.0,inf,0.0,0.5,0.0,0.75
3,(A),(C),1.0,0.5,0.5,0.5,1.0,1.0,0.0,1.0,0.0,0.5,0.0,0.75
4,(B),(C),1.0,0.5,0.5,0.5,1.0,1.0,0.0,1.0,0.0,0.5,0.0,0.75
5,(C),(B),0.5,1.0,0.5,1.0,1.0,1.0,0.0,inf,0.0,0.5,0.0,0.75
6,"(B, C)",(A),0.5,1.0,0.5,1.0,1.0,1.0,0.0,inf,0.0,0.5,0.0,0.75
7,"(B, A)",(C),1.0,0.5,0.5,0.5,1.0,1.0,0.0,1.0,0.0,0.5,0.0,0.75
8,"(C, A)",(B),0.5,1.0,0.5,1.0,1.0,1.0,0.0,inf,0.0,0.5,0.0,0.75
9,(B),"(C, A)",1.0,0.5,0.5,0.5,1.0,1.0,0.0,1.0,0.0,0.5,0.0,0.75


## TUGAS

### Soal 1
**Unduh Dataset**

Gunakan dataset Online Retail Dataset dari UCI Machine Learning Repository. Dataset ini berisi transaksi ritel online, termasuk informasi tentang ID produk, jumlah, dan negara pelanggan.

Link dataset: https://archive.ics.uci.edu/ml/datasets/Online+Retail

Pertama-tama saya terlebih dahulu mengunduh dataset (file excel) kemudian mengupload ke GitHub untuk kemudian mengambil link raw dataset

Selanjutnya mengimport dataset dan memasukkannya kedalam dataframe pandas

In [30]:
import pandas as pd

# Url dataset via GitHub
url = 'https://github.com/helloohowareyu/Praktikum-Kecerdasan-Artifisial-Lanjut/raw/refs/heads/main/Online%20Retail.xlsx'

df = pd.read_excel(url)

### Soal 2
**Eksplorasi Data**

In [31]:
# Menampilkan 5 entri pertama dari dataset
df.head(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


Kolom yang relevan untuk diidentifikasi adalah InvoiceNO yang digunakan sebagai kunci unik transaksi untuk mengelompokkan item yang dibeli bersamaan (basket) dan Description sebagai variabel item yang akan dianalisis hubungannya dalam aturan asosiasi

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


Karena jumlah entri yang sangat besar (sektiar 541 ribu), maka akan dilakukan pemotongan jumlah entri

Alih-alih memotong data secara acak (yang bisa merusak pola belanja), cara yang saya gunakan adalah mengambil entri dari satu negara saja

In [32]:
# Menghitung jumlah transaksi unik (InvoiceNo) per negara
country_counts = df.groupby('Country')['InvoiceNo'].nunique().sort_values(ascending=False)

# Menampilkan hasil
country_counts

,InvoiceNo
Country,
United Kingdom,23494
Germany,603
France,461
EIRE,360
Belgium,119
Spain,105
Netherlands,101
Switzerland,74
Portugal,71


Terlihat mayoritas data didapat dari negara UK dan entrinya adalah yang paling banyak, untuk itu kita gunakan negara lain dengan jumlah terbanyak kedua yaitu Jerman dengan jumlah entrinya adalah 603 yang cukup untuk digunakan di tugas kali ini.

In [33]:
# Filter dataset untuk negara 'Germany'
df_germany = df[df['Country'] == 'Germany'].copy()

Selanjutnya kita akan mulai dengan membersihkan entri dengan menghapus transaksi dengan InvoiceNo yang kosong atau Nan dan menghapus transaksi yang dibatalkan (dimulai dengan 'C')

In [34]:
df_germany.dropna(subset=['InvoiceNo', 'Description'], inplace=True)

df_germany = df_germany[~df_germany['InvoiceNo'].astype(str).str.contains('C')]

Selanjutnya adalah menghapus entri dengan kuantitas kurang dari atau sama dengan 0, yang biasanya adalah return barang. Selain itu kita akan bersihkan spasi berlebih di deskripsi produk

In [35]:
df_germany = df_germany[df_germany['Quantity'] > 0]

df_germany['Description'] = df_germany['Description'].str.strip()

In [37]:
df_germany.head(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
1109,536527,22809,SET OF 6 T-LIGHTS SANTA,6,2010-12-01 13:04:00,2.95,12662.0,Germany
1110,536527,84347,ROTATING SILVER ANGELS T-LIGHT HLDR,6,2010-12-01 13:04:00,2.55,12662.0,Germany
1111,536527,84945,MULTI COLOUR SILVER T-LIGHT HOLDER,12,2010-12-01 13:04:00,0.85,12662.0,Germany
1112,536527,22242,5 HOOK HANGER MAGIC TOADSTOOL,12,2010-12-01 13:04:00,1.65,12662.0,Germany
1113,536527,22244,3 HOOK HANGER MAGIC GARDEN,12,2010-12-01 13:04:00,1.95,12662.0,Germany


In [38]:
df_germany.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9042 entries, 1109 to 541805
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   InvoiceNo    9042 non-null   object        
 1   StockCode    9042 non-null   object        
 2   Description  9042 non-null   object        
 3   Quantity     9042 non-null   int64         
 4   InvoiceDate  9042 non-null   datetime64[ns]
 5   UnitPrice    9042 non-null   float64       
 6   CustomerID   9042 non-null   float64       
 7   Country      9042 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 893.8+ KB


### Soal 3
**Persiapan Data untuk Association Rule**

Di persiapan data, kita akan ubah data transaksi menjadi matriks dimana baris adalah ID transaksi dan kolom adalah item menggunakan pd.crosstab dan applymap untuk mengubah nilai (kuantitas) menjadi 1 jika barang tersebut dibeli dan 0 jika tidak

In [43]:
basket = pd.crosstab(index=df_germany['InvoiceNo'], columns=df_germany['Description']).map(lambda x: 1 if x > 0 else 0)
basket

Description,10 COLOUR SPACEBOY PEN,12 COLOURED PARTY BALLOONS,12 IVORY ROSE PEG PLACE SETTINGS,12 MESSAGE CARDS WITH ENVELOPES,12 PENCIL SMALL TUBE WOODLAND,12 PENCILS SMALL TUBE RED RETROSPOT,12 PENCILS SMALL TUBE SKULL,12 PENCILS TALL TUBE POSY,12 PENCILS TALL TUBE RED RETROSPOT,12 PENCILS TALL TUBE SKULLS,...,YULETIDE IMAGES GIFT WRAP SET,ZINC HEART T-LIGHT HOLDER,ZINC STAR T-LIGHT HOLDER,ZINC BOX SIGN HOME,ZINC FOLKART SLEIGH BELLS,ZINC HEART LATTICE T-LIGHT HOLDER,ZINC METAL HEART DECORATION,ZINC T-LIGHT HOLDER STAR LARGE,ZINC T-LIGHT HOLDER STARS SMALL,ZINC WILLIE WINKIE CANDLE STICK
InvoiceNo,,,,,,,,,,,,,,,,,,,,,
536527,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536840,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536861,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536967,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536983,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
581266,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
581494,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
581570,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


### Soal 4
**Penerapan Algoritma**

Kita akan terapkan algoritma apriori untuk mencari kombinasi item yang paling sering muncul sesering nilai `min_support`.

`min_support` yang kita gunakan kali ini adalah 0.02, artinya item harus muncul minimal di 2% dari seluruh transaksi.

In [47]:
frequent_itemsets = apriori(basket, min_support=0.02, use_colnames=True)
frequent_itemsets

,support,itemsets
0,0.024070,(10 COLOUR SPACEBOY PEN)
1,0.021882,(12 PENCIL SMALL TUBE WOODLAND)
2,0.032823,(3 HOOK HANGER MAGIC GARDEN)
3,0.041575,(3 PIECE SPACEBOY COOKIE CUTTER SET)
4,0.024070,(36 PENCILS TUBE RED RETROSPOT)
...,...,...
910,0.021882,"(POSTAGE, ROUND SNACK BOXES SET OF 4 FRUITS, R..."
911,0.035011,"(POSTAGE, ROUND SNACK BOXES SET OF 4 FRUITS, S..."
912,0.024070,"(STRAWBERRY LUNCH BOX WITH CUTLERY, POSTAGE, R..."
913,0.035011,"(WOODLAND CHARLOTTE BAG, POSTAGE, ROUND SNACK ..."


### Soal 5
**Evaluasi Aturan Asosiasi**

Setelah menerapkan apriori, kita akan menghasilkan aturan asosiasi dari `frequent_itemsets`

In [51]:
rules = association_rules(frequent_itemsets, metric='lift', min_threshold=1)

Selanjutnya kita akan ambil 3 aturan terbaik berdasarkan nilai lift

In [54]:
top_rules = rules.nlargest(3, 'lift')
top_rules

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
208,(DOLLY GIRL CHILDRENS CUP),(DOLLY GIRL CHILDRENS BOWL),0.024070,0.026258,0.024070,1.000000,38.083333,1.0,0.023438,inf,0.997758,0.916667,1.000000,0.958333
209,(DOLLY GIRL CHILDRENS BOWL),(DOLLY GIRL CHILDRENS CUP),0.026258,0.024070,0.024070,0.916667,38.083333,1.0,0.023438,11.71116,1.000000,0.916667,0.914611,0.958333
1000,"(DOLLY GIRL CHILDRENS CUP, POSTAGE)",(DOLLY GIRL CHILDRENS BOWL),0.021882,0.026258,0.021882,1.000000,38.083333,1.0,0.021307,inf,0.995526,0.833333,1.000000,0.916667


Dari hasil penerapan rules kita dapatkan:
1. Aturan [DOLLY GIRL CHILDRENS CUP] $\rightarrow$ [DOLLY GIRL CHILDRENS BOWL]: Memiliki nilai lift sebesar 38.08, yang menunjukkan korelasi positif yang kuat antara kedua produk tersebut. Dengan tingkat confidence sebesar 1
2. Aturan [DOLLY GIRL CHILDRENS BOWL] $\rightarrow$ [DOLLY GIRL CHILDRENS CUP]: Memiliki nilai lift sebesar 38.08, yang menunjukkan korelasi positif yang kuat antara kedua produk tersebut. Dengan tingkat confidence sebesar 0.9
3. Aturan [DOLLY GIRL CHILDRENS CUP, POSTAGE] $\rightarrow$ [DOLLY GIRL CHILDRENS BOWL]: Memiliki nilai lift sebesar 38.08, yang menunjukkan korelasi positif yang kuat antara kedua produk tersebut. Dengan tingkat confidence sebesar 1

### Kesimpulan

Berdasarkan praktikum yang telah dilakukan menggunakan **Online Retail Dataset** (fokus pada negara Jerman), berikut adalah kesimpulan yang dapat ditarik:

1.  **Pencapaian Tujuan Percobaan:**
    Tujuan untuk mengimplementasikan algoritma **Apriori** dan **Association Rules** telah tercapai. Kita berhasil mengidentifikasi pola pembelian pelanggan dan menghasilkan aturan asosiasi yang signifikan berdasarkan nilai *support*, *confidence*, dan *lift*.

2.  **Hubungan Hasil dengan Dasar Teori:**
    Hasil analisis menunjukkan adanya hubungan fungsional antar produk. Misalnya, aturan antara `DOLLY GIRL CHILDRENS CUP` dan `DOLLY GIRL CHILDRENS BOWL` memiliki nilai **Lift (38.08)** yang sangat tinggi, yang secara teori membuktikan adanya korelasi positif yang kuat (jauh di atas 1). Hal ini memvalidasi teori bahwa item yang memiliki kegunaan serupa seringkali dibeli bersamaan dalam satu transaksi.

3.  **Faktor yang Memengaruhi Hasil:**
    - **Pembersihan Data:** Penghapusan transaksi yang dibatalkan ('C') dan kuantitas negatif sangat krusial agar pola yang ditemukan mencerminkan perilaku belanja yang valid.
    - **Nilai Ambang Batas (Threshold):** Penentuan `min_support=0.02` memungkinkan kita menemukan itemset yang cukup representatif tanpa membebani komputasi, sementara `min_threshold=1` pada *lift* memastikan hanya aturan dengan korelasi positif yang ditampilkan.
    - **Seleksi Data:** Pemfokusan pada satu negara (Jerman) membantu dalam mendapatkan pola yang lebih homogen dibandingkan jika data dari seluruh dunia digabungkan secara acak.

4.  **Temuan Utama:**
    Strategi penempatan barang atau promo paket (bundling) dapat difokuskan pada kategori produk anak-anak (seperti seri Dolly Girl) karena memiliki tingkat kepercayaan (*confidence*) yang mencapai 0.9 hingga 1.0.